### PR TITLE
Add GraphPath struct in Runtime

### PR DESCRIPTION
**Pull request by BingqingLyu**
## What do these changes do?

1. Add GraphPath struct in Runtime to support path_expand.
2. Refine OnGraph RecordElement, i.e., make it consist of Vertex, Edge, and GraphPath.
3. Simplify PartialEq, PartialOrd, Hash trait implementation for Record etc.

## Related issue number

Fixes

### CODE REVIEW

File: research/query_service/ir/runtime/src/graph/element/mod.rs [hunk 1]
```diff
@@ -204,3 +207,109 @@ impl TryFrom<result_pb::Edge> for VertexOrEdge {
...
+
+impl Hash for VertexOrEdge {
+    fn hash<H: Hasher>(&self, mut state: &mut H) {
+        self.id().hash(&mut state)
+    }
+}
+
+#[derive(Clone, Debug, Hash)]
+pub enum GraphObject {
+    VOrE(VertexOrEdge),
```
Comment (longbinlai) [inline research/query_service/ir/runtime/src/graph/element/mod.rs:219]: 话说，这个为啥不用Vertex, Edge，然后VertexOrEdge就可以成为历史了

Comment (BingqingLyu) [reply research/query_service/ir/runtime/src/graph/element/mod.rs:219]: 保留VertexOrEdge是出于，抛开上层表示不说，其实path是可以 点、边、点、边这样的，所以保留了这个

Comment (BingqingLyu) [reply research/query_service/ir/runtime/src/graph/element/mod.rs]: 我改了GraphObject结构，目前是这样，GraphObject包含Vertex, Edge, GraphPath，暴露给Record Entry使用。另外我在GraphPath里额外定义了VertexOrEdge，这个只给GraphPath使用，使得GraphPath能够成为一条能有点和边的path.

File: research/query_service/ir/runtime/src/graph/element/path.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    fn as_borrow_object(&self) -> BorrowObject {
+        BorrowObject::None
+    }
+}
+
+impl GraphElement for GraphPath {
+    fn id(&self) -> ID {
+        todo!()
```
Comment (longbinlai) [inline research/query_service/ir/runtime/src/graph/element/path.rs:96]: can hash() all the vertex ids in the path.

File: research/query_service/ir/runtime/src/graph/element/path.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    fn label(&self) -> Option<&NameOrId> {
```
Comment (longbinlai) [inline research/query_service/ir/runtime/src/graph/element/path.rs:99]: Can concatenate all labels in the paths

File: research/query_service/ir/runtime/src/graph/element/path.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+#[derive(Clone, Debug, Hash, PartialEq, PartialOrd)]
+pub enum PathStruct {
+    // save the whole path
+    WHOLE(Vec<VertexOrEdge>),
+    // save the path_end
+    END(VertexOrEdge),
+}
+
+#[derive(Clone, Debug, Hash)]
+pub struct GraphPath {
```
Comment (longbinlai) [inline research/query_service/ir/runtime/src/graph/element/path.rs:53]: Why not merge GraphPath and PathStruct?

File: research/query_service/ir/runtime/src/graph/element/path.rs (hunk 2 above)
Comment (longbinlai) [reply research/query_service/ir/runtime/src/graph/element/path.rs:96]: all ids()

Comment (BingqingLyu) [reply research/query_service/ir/runtime/src/graph/element/path.rs]: Done

File: research/query_service/ir/runtime/src/graph/element/path.rs (hunk 4 above)
Comment (BingqingLyu) [reply research/query_service/ir/runtime/src/graph/element/path.rs]: Done

File: research/query_service/ir/runtime/src/graph/element/path.rs (hunk 3 above)
Comment (BingqingLyu) [reply research/query_service/ir/runtime/src/graph/element/path.rs:156]: Seems that I cannot create a String and return it as Option<&NameOrId::Str> by ref in label(). So I leave it as return None for now.

Comment (longbinlai) [reply research/query_service/ir/runtime/src/graph/element/path.rs:156]: Fine

**Pull request merged**